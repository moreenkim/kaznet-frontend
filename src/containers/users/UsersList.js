import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import Moment from "react-moment";
import queryString from "query-string";

import * as userActions from "../../store/users/actions";
import * as globalActions from "../../store/global/actions";
import * as userSelectors from "../../store/users/reducer";
import * as constants from "../../constants.js";

import ListView from "../../components/ListView";
import ElementMap from "../ElementMap";

export class UsersList extends Component {
  async componentDidMount() {
    this.props.showListTitle();
    this.props.changePageTitle("Users");
    this.props.changePageTitleButton("+ Create User");

    if (/\?page=(\d|\w)/.test(this.props.location.search)) {
      const { page } = queryString.parse(this.props.location.search);
      const pageNumber = Number(page);
      if (!isNaN(pageNumber)) {
        await this.props.fetchUsers(
          `${constants.API_ENDPOINT}/userprofiles/?page=${pageNumber}`
        );
        this.props.changePageNumber(pageNumber);
      }
    } else {
      this.props.fetchUsers();
    }
  }

  componentDidUpdate(prevProps) {
    if (/\?page=(\d|\w)/.test(this.props.location.search)) {
      const { page } = queryString.parse(this.props.location.search);
      if (Number(page) !== this.props.currentPage && !isNaN(page)) {
        const pageNumber = Number(page);
        this.props.fetchUsers(
          `${constants.API_ENDPOINT}/userprofiles/?page=${pageNumber}`
        );
        this.props.changePageNumber(pageNumber);
      }
    }
  }

  render() {
    if (!this.props.rowsById) return this.renderLoading();
    return (
      <div className="UsersList">
        <ListView
          renderHeaders={this.renderHeaders}
          rowsIdArray={this.props.rowsIdArray}
          rowsById={this.props.rowsById}
          renderRow={this.renderRow}
          endpoint={"users"}
          pageLinks={this.props.pageLinks}
          totalPages={this.props.totalPages}
          currentPage={this.props.currentPage}
          firstPage={this.props.firstPage}
          lastPage={this.props.lastPage}
        />
      </div>
    );
  }

  renderLoading() {
    return <p>Loading...</p>;
  }

  renderHeaders() {
    const headerItems = [
      "Role",
      "Username",
      "Last Name",
      "First Name",
      "Submissions",
      "Approved %",
      "Last Active Date"
    ];
    return <ElementMap items={headerItems} HTMLTag="th" />;
  }

  renderRow(row) {
    const rowItems = [
      row.attributes.role_display,
      row.attributes.ona_username,
      row.attributes.last_name,
      row.attributes.first_name,
      row.attributes.submission_count,
      row.attributes.approval_rate * 100 + "%",
      <Moment key={row.id} format="DD-MM-YYYY">
        {row.attributes.last_login}
      </Moment>
    ];

    return <ElementMap items={rowItems} HTMLTag="td" />;
  }
}

// which props do we want to inject, given the global store state?
function mapStateToProps(state) {
  return {
    rowsById: userSelectors.getUsersById(state),
    rowsIdArray: userSelectors.getUsersIdArray(state),
    totalPages: userSelectors.getTotalPages(state),
    currentPage: userSelectors.getCurrentPage(state),
    pageLinks: userSelectors.getPageLinks(state),
    firstPage: userSelectors.getFirstPage(state),
    lastPage: userSelectors.getLastPage(state)
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchUsers: userActions.fetchUsers,
      changePageNumber: userActions.changePageNumber,
      changePageTitle: globalActions.changePageTitle,
      changePageTitleButton: globalActions.changePageTitleButton,
      showListTitle: globalActions.toggleDetailTitleOff
    },
    dispatch
  );
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UsersList);
