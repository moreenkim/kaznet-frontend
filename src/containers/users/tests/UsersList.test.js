import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";

import { UsersList } from "../UsersList";
import * as fixtures from "../../../store/users/tests/fixtures";

const history = createBrowserHistory();

describe("containers/users/UsersList", () => {
  it("renders without crashing", () => {
    shallow(
      <UsersList
        fetchUsers={function() {}}
        changePageTitle={function() {}}
        changePageTitleButton={function() {}}
        showListTitle={function() {}}
      />
    );
  });

  it("renders user list correctly", () => {
    const wrapper = mount(
      <Router history={history}>
        <UsersList
          fetchUsers={function() {}}
          changePageTitle={function() {}}
          changePageTitleButton={function() {}}
          showListTitle={function() {}}
          rowsById={fixtures.usersById}
          rowsIdArray={fixtures.usersIdArray}
          endpoint={"userprofiles"}
          pageLinks={fixtures.pageLinks}
          totalPages={fixtures.totalPages}
          currentPage={fixtures.currentPage}
          firstPage={fixtures.firstPage}
          lastPage={fixtures.lastPage}
        />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
