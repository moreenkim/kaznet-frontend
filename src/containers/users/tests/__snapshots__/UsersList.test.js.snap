// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/users/UsersList renders user list correctly 1`] = `
<UsersList
  changePageTitle={[Function]}
  changePageTitleButton={[Function]}
  fetchUsers={[Function]}
  rowsById={
    Object {
      "1": Object {
        "attributes": Object {
          "amount_earned": "0 KES",
          "approval_rate": 0,
          "approved_submissions": 0,
          "avg_amount_earned": 0,
          "avg_approval_rate": 0,
          "avg_approved_submissions": 0,
          "avg_rejected_submissions": 0,
          "avg_submissions": 0,
          "created": "2018-06-19T16:03:10.184043+03:00",
          "email": "sol@admin.me",
          "expertise": "1",
          "first_name": "Davis",
          "gender": "1",
          "last_login": "2018-06-19T16:03:22+03:00",
          "last_name": "Raymond",
          "modified": "2018-06-20T08:45:12.976916+03:00",
          "national_id": null,
          "ona_pk": null,
          "ona_username": "davisraym",
          "payment_number": "",
          "phone_number": "",
          "rejected_submissions": 0,
          "role": "1",
          "role_display": "Admin",
          "submission_count": 0,
        },
        "id": "1",
        "type": "UserProfile",
      },
    }
  }
  rowsIdArray={
    Array [
      "1",
    ]
  }
  showListTitle={[Function]}
>
  <div
    className="UsersList"
  >
    <ListView
      renderHeaders={[Function]}
      renderRow={[Function]}
      rowsById={
        Object {
          "1": Object {
            "attributes": Object {
              "amount_earned": "0 KES",
              "approval_rate": 0,
              "approved_submissions": 0,
              "avg_amount_earned": 0,
              "avg_approval_rate": 0,
              "avg_approved_submissions": 0,
              "avg_rejected_submissions": 0,
              "avg_submissions": 0,
              "created": "2018-06-19T16:03:10.184043+03:00",
              "email": "sol@admin.me",
              "expertise": "1",
              "first_name": "Davis",
              "gender": "1",
              "last_login": "2018-06-19T16:03:22+03:00",
              "last_name": "Raymond",
              "modified": "2018-06-20T08:45:12.976916+03:00",
              "national_id": null,
              "ona_pk": null,
              "ona_username": "davisraym",
              "payment_number": "",
              "phone_number": "",
              "rejected_submissions": 0,
              "role": "1",
              "role_display": "Admin",
              "submission_count": 0,
            },
            "id": "1",
            "type": "UserProfile",
          },
        }
      }
      rowsIdArray={
        Array [
          "1",
        ]
      }
    >
      <Table
        bordered={true}
        className="kaznet-table"
        responsiveTag="div"
        tag="table"
      >
        <table
          className="kaznet-table table table-bordered"
        >
          <thead>
            <tr>
              <ElementMap
                HTMLTag="th"
                items={
                  Array [
                    "Role",
                    "Username",
                    "Last Name",
                    "First Name",
                    "Submissions",
                    "Approved %",
                    "Last Active Date",
                  ]
                }
              >
                <th
                  key="0"
                >
                  Role
                </th>
                <th
                  key="1"
                >
                  Username
                </th>
                <th
                  key="2"
                >
                  Last Name
                </th>
                <th
                  key="3"
                >
                  First Name
                </th>
                <th
                  key="4"
                >
                  Submissions
                </th>
                <th
                  key="5"
                >
                  Approved %
                </th>
                <th
                  key="6"
                >
                  Last Active Date
                </th>
              </ElementMap>
            </tr>
          </thead>
          <tbody>
            <tr
              key="1"
            >
              <ElementMap
                HTMLTag="td"
                items={
                  Array [
                    "Admin",
                    "davisraym",
                    "Raymond",
                    "Davis",
                    0,
                    "0%",
                    <t
                      ago={false}
                      calendar={false}
                      decimal={false}
                      element="time"
                      filter={[Function]}
                      format="DD-MM-YYYY"
                      fromNow={false}
                      interval={60000}
                      onChange={[Function]}
                      toNow={false}
                      unit={null}
                      unix={false}
                      utc={false}
                  >
                      2018-06-19T16:03:22+03:00
                  </t>,
                  ]
                }
              >
                <td
                  key="0"
                >
                  Admin
                </td>
                <td
                  key="1"
                >
                  davisraym
                </td>
                <td
                  key="2"
                >
                  Raymond
                </td>
                <td
                  key="3"
                >
                  Davis
                </td>
                <td
                  key="4"
                >
                  0
                </td>
                <td
                  key="5"
                >
                  0%
                </td>
                <td
                  key="6"
                >
                  <t
                    ago={false}
                    calendar={false}
                    decimal={false}
                    element="time"
                    filter={[Function]}
                    format="DD-MM-YYYY"
                    fromNow={false}
                    interval={60000}
                    key="1"
                    onChange={[Function]}
                    toNow={false}
                    unit={null}
                    unix={false}
                    utc={false}
                  >
                    <time
                      dateTime={"2018-06-19T13:03:22.000Z"}
                    >
                      19-06-2018
                    </time>
                  </t>
                </td>
              </ElementMap>
            </tr>
          </tbody>
        </table>
      </Table>
    </ListView>
  </div>
</UsersList>
`;
